### ISSUE TITLE
Bindgen generates suboptimal constants for unnamed enums with fixed underlying types

### ISSUE THREAD

**Issue by csmulhern** [rust-lang/rust-bindgen#1828]
### Input C/C++ Header

```C++
typedef char RSType;

enum : RSType {
    kRSTypeOne,
    kRSTypeTwo,
    kRSTypeThree
};
```

### Bindgen Invocation

```
$ bindgen bindgen.h -o bindings.rs --no-layout-tests -- -x c++ -std=c++11
```

### Actual Results

```rust
/* automatically generated by rust-bindgen 0.54.1 */

pub type RSType = ::std::os::raw::c_char;
pub const kRSTypeOne: _bindgen_ty_1 = 0;
pub const kRSTypeTwo: _bindgen_ty_1 = 1;
pub const kRSTypeThree: _bindgen_ty_1 = 2;
pub type _bindgen_ty_1 = i8;
```

### Expected Results

```rust

pub type RSType = ::std::os::raw::c_char;
pub const kRSTypeOne: RSType = 0;
pub const kRSTypeTwo: RSType = 1;
pub const kRSTypeThree: RSType = 2;
```

For context, this issue is common in my usage of bindgen to generate bindings for macOS frameworks, which commonly use the CF_ENUM macro defined in <CoreFoundation/CFAvailability.h>. This issue only occurs when the unnamed variant is used, causing the enum definition to be anonymous. I'm guessing it's related to the bindgen typename being {enum name}_{variant name}, but there being no enum name in this case. As an aside, is the naming convention described somewhere, and is it controllable? I find it a little odd that the rust constant name uses the enum name as a prefix when the symbol in C is not prefixed (and is unique). I ask because it's simplest for me if the implementation code wrapping these bindings can use the same symbol names as the equivalent C code. I'm curious if the prefixing I'm describing can be disabled, or if the project would be open to accepting a PR that adds the ability to do so.

**Comment by csmulhern** [comment]
I spent a little of time looking into this, and discovered that _bindgen_ty_1 is the canonical name of the anonymous enum. As such, I think this expected output would be more in line with the named enum case:

```
/* automatically generated by rust-bindgen 0.54.1 */

pub type RSType = ::std::os::raw::c_char;
pub const kRSTypeOne: _bindgen_ty_1 = 0;
pub const kRSTypeTwo: _bindgen_ty_1 = 1;
pub const kRSTypeThree: _bindgen_ty_1 = 2;
pub type _bindgen_ty_1 = RSType;
```

I also discovered that --no-prepend-enum-name solves the other issue I brought up.

**Comment by emilio** [comment]
Sorry for the lag in getting to this, looking now :)

**Comment by emilio** [comment]
So... right, this is somewhat unfortunate because anonymous enums are somewhat hard to handle, specially given all the different tweaks that you can apply in bindgen to get different APIs / generated code for enums.

We need the underlying integer type to get the right layout for some of those. We lose that information in the `canonical_type` call right below the code you linked... But I think we could preserve that type a bit further without too much trouble I think.

So for example something like this:

```diff
diff --git a/src/codegen/mod.rs b/src/codegen/mod.rs
index c68eb6da..d0b96221 100644
--- a/src/codegen/mod.rs
+++ b/src/codegen/mod.rs
@@ -2859,9 +2859,12 @@ impl CodeGenerator for Enum {
             });
         }
 
-        let repr = {
-            let repr_name = ctx.rust_ident_raw(repr_name);
-            quote! { #repr_name }
+        let repr = match self.repr() {
+            Some(ty) => ty.to_rust_ty_or_opaque(ctx, &()),
+            None => {
+                let repr_name = ctx.rust_ident_raw(repr_name);
+                quote! { #repr_name }
+            },
         };
 
         let mut builder = EnumBuilder::new(
```

Gets us the first expected output in your last comment. Then, for the case of an unnamed enum and the `Consts` enum style, we could go further and use that `repr` as the type of the different constants directly, which would get the best output I think. Then we could even omit the `_bindgen_ty_1` definition altogether in that case, if wanted.

I'd be fancy with any of those options getting submitted as a patch along with some tests :)

**Comment by emilio** [comment]
(Renaming the issue title to clarify that I don't think the types are _technically_ incorrect, unless I'm missing something)

**Comment by emilio** [comment]
Fixed by #1850, thanks @csmulhern!
